### ISSUE TITLE
Follow up of #160, mapping method selection should select IterableMapping and MapMapping as well.

### ISSUE THREAD

**Issue by sjaakd** [mapstruct/mapstruct#487]
Allow update method selection mechanism for properties to work on collections and maps.

**Comment by sjaakd** [comment]
There are several cases to consider here:
1. Lists
2. Sets
3. Arrays
4. Maps.

Lets start with 1. 

Assume we have a
- source `List` with elements `[A, C, D]` 
- target `List` with elements `[A', B' D']`
- there is a mapping method (both create and update)  available for the elements `X -> X`]

I would expect the result to look like:
`[A'', C' D'']`
Where: 
- `A'` and `D'` are updated to `A''` respectively `D''`
- `B'` is deleted, 
- `C'` is created

All in all the order is maintained.

For 2. Its likewise, only the order is irrelevant.

For 3. The array size needs to be the same. Any deviation of order cannot be mapped and should throw an exception (which?)

For 4. The should be the determining factor. In absence of a key in the target, a new entry will be added. Superflous keys will be removed.

This will all depends on meaningful `hashcode` and `equals` methods in the target objects. 

-------------------
Solution wise:

We also need some intermediate results. The source collection / map first needs to be mapped completely (we could call a forged method for that if a mapping is absent). And then we execute the merging logic along the lines that is described above.

**Comment by nocquidant** [comment]
FYI GeDa uses matcher objects.
http://www.inspire-software.com/confluence/display/GeDA/Annotations+API#AnnotationsAPI-@DtoCollection.dtoToEntityMatcher

But the O(n^2) could be improved I guess.

**Comment by sjaakd** [comment]
> FYI GeDa uses matcher objects

Lets first try to tackle the simple case. Would you be interested to help out on this one?

**Comment by nocquidant** [comment]
I'm afraid I do not find the time to do it. Hats off ;)

**Comment by sjaakd** [comment]
As an in-between solution we could consider selection of handwritten methods with target type collection / map as a first step.

**Comment by agudian** [comment]
@sjaakd, we should close this, right?

**Comment by sjaakd** [comment]
Nope.. This one concerns also generation of these mappings. So this is only the first (small) step.

**Comment by agudian** [comment]
Ok, then I'd suggest setting the milestone to 1.1... :)

**Comment by sjaakd** [comment]
> Ok, then I'd suggest setting the milestone to 1.1... :)

ack.

**Comment by marcoschulte** [comment]
We might find the time to implement that feature.

I guess it should be discussed beforehand how that problem could/should be solved.

I'm not aware about the terms you are using, therefore I clarify what I mean by _mapping_ and _merging_:

To me a mapping method looks like `B map(A source)` and a merging function like `void merge(A source, @MappingTarget B target)`

Basically I see two options to get the collection mapping right:

1. Implement a merge logic like you mentioned in https://github.com/mapstruct/mapstruct/issues/487#issuecomment-87619152. That would make it necessary to create the aforementioned matchers to find the target instance in the list.
2. Extend the object factory concept and always clear collections: If the objectmapper would also be supplied with the source object in addition to the target class, it could decide whether it should load the entity via the entitiymanager (if an id is set in the source object) or call the default constructor on the target class (if id is still null). To me this seems to be a quite elegant solution, as this should work both with mapping and merging, and should solve the O(n^2) problem - the list entities should be cached in the 2nd level cache anyways and they are ideally loaded via the technical id by the object factory. To work with _mapping_ methods and nested collections one just needs to make sure that `clear()` is used everywhere instead of instantiating new collections. IMO OneToOne relations also work with that approach straight away.

What could also be discussed is, if this logic should only be implemented for merging. This would require to stay in the "merging" logic when traversing down, which I already described in https://github.com/mapstruct/mapstruct/issues/153#issuecomment-241717848 and is currently not the case. IMO this should be usable for both for mapping and merging.

In fact we had the second approach already working for collections. It just broke at the second level of nested lists, as the mapper wasn't in the _merging_ branch anymore and was replacing the collections instead of clearing them:

```java
@Component
public class EntityRepository {

	@PersistenceContext
	private EntityManager em;

	private ThreadLocal<Long> idStore = new ThreadLocal<>();

	@BeforeMapping
	public <T extends TransientModel, E extends PersistentModel> void before(T transientModel, @TargetType Class<E>
			clazz) {
		idStore.set(transientModel.getId());
	}

	public <T extends PersistentModel> T loadById(@TargetType Class<T> clazz) {
		Long id = idStore.get();

		if (id == null) {
			try {
				return clazz.newInstance();
			} catch (InstantiationException | IllegalAccessException e) {
				throw new RuntimeException(e);
			}
		}

		T result = em.find(clazz, id);
		Assert.notNull(result, "Invalid id " + id + " for entity " + clazz);

		return result;
	}
}
```

**Comment by agudian** [comment]
@marcoschulte that'd be great!

Some remarks: what you call _merging_ methods is what ended up being called _update_ methods in our issues, docs, and partially in the code and tests - in the code you can also find the term `isExistingInstanceMapping` in some places instead of `isUpdate` or something -- that might be something to refactor properly, I guess.

As for the factory methods, we have that on the radar and are tracking it in this issue: #824. Interesting workaround you're having there, though... :wink: 

Quite some time went by with this issue not being touched, so I think we should really go for the simple solution, which is what you're describing as option 2: hand down the collection (or map -- this should cover maps as well) to the iterable-update method. When generating an iterable mapping method with an existing collection, we already add a `clear()` call within the iterable-mapping method before adding newly mapped items. I think we should leave the `clear()` call at that place, as it makes it possible to define hand-written update-methods that do things differently (e.g. actually _merging_ the lists with updating existing instances and identifying them with matchers).
Adding MapStruct-support for smarter merge logic with user-specified matchers could follow up later when necessary without having to break the backwards-compatibility.

**Comment by marcoschulte** [comment]
To sum it up:
* #824 should be implemented. As far as I understand it, you decided to use a method level annotation called `@TargetBeanFactory` or `@ObjectFactory`. Should backwards compatibility be broken and the annotation become necessary for all factory methods, or is the annotation optional?
* I haven't had a look at the mapstruct sources, but I assume you are talking about an iterable-update method that is already existing and is chosen in update methods? This method stays untouched with regards to clearing the collections. There is probably another method a la iterable-create-new-list which is chosen in map methods? I assume the latter one can be removed to fix the problem with nested collections?

**Comment by dsrini-open** [comment]
Any plans when this might/will/ be implemented? Or is it implemented? 
I am looking for an answer to this [Question](https://stackoverflow.com/questions/62900571/use-update-mehtod-mapping-from-other-mapper)
